### PR TITLE
fix: bugs when serialize union type

### PR DESCRIPTION
**Pull request by maxwelljin**
This bug is caused by the deserialization process. The deserialization function currently only checks the schema of the class, without specific information about the particular document type in the field of union type. To resolve this issue, we may need to modify the data storage. After discussing with Joan, we could temporarily remove support for the union type and add this info the document. Additionally, we provide users with a more informative error message.

### CODE REVIEW

File: docs/user_guide/sending/serialization.md [hunk 1]
```diff
@@ -303,4 +303,6 @@ assert dv_from_proto_numpy.tensor_type == NdArray
 assert isinstance(dv_from_proto_numpy.tensor, NdArray)
 ```
 
+!!! note
+    Currently, union type hints are not supported for serialization to protobuf formats.
```
Comment (JoanFM) [inline docs/user_guide/sending/serialization.md:307]: Any union type? or Union types of regular types are? Can u add tests for the exceptions?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: docs/user_guide/sending/serialization.md (hunk 1 above)
Comment (maxwelljin) [reply docs/user_guide/sending/serialization.md:307]: I'll add more tests. It should be any union type as union type doesn't have the `from_protobuf` method

**Comment by maxwelljin** [comment]
Currently, the serialization for union type is only supported through `from_json` and `to_json`. This new commit adds tests to this bug.

File: docs/user_guide/sending/serialization.md [hunk 2]
```diff
@@ -304,5 +304,5 @@ assert isinstance(dv_from_proto_numpy.tensor, NdArray)
 ```
 
 !!! note
-    Currently, union type hints are not supported for serialization to protobuf formats.
+    Serialization to protobuf formats is not supported for union type hints. Please consider using JSON serialization as an alternative.
```
Comment (JoanFM) [inline docs/user_guide/sending/serialization.md:307]: Avoid the consideration, just inform

File: tests/units/array/test_array_from_to_bytes.py [hunk 3]
```diff
@@ -69,3 +69,17 @@ def test_from_to_base64(protocol, compress, show_progress):
...
     assert da[1].image.url is None
     assert da2[1].image.url is None
+
+
+def test_union_type_error(tmp_path):
+    from typing import Union
+
+    from docarray.documents import TextDoc
+
+    class CustomDoc(BaseDoc):
```
Comment (JoanFM) [inline tests/units/array/test_array_from_to_bytes.py:79]: what about Union[int, float] or Union[int, str]

**Review by JoanFM — CHANGES_REQUESTED** [review]
Still not answered what happens with `Union` of basic types

File: tests/units/array/test_array_from_to_bytes.py (hunk 3 above)
Comment (maxwelljin) [reply tests/units/array/test_array_from_to_bytes.py:79]: After the testing, the Union with basis types should work in all serialization methods. Only the Union with document types will cause those issues.

File: docs/user_guide/sending/serialization.md [hunk 4]
```diff
@@ -303,4 +303,8 @@ assert dv_from_proto_numpy.tensor_type == NdArray
 assert isinstance(dv_from_proto_numpy.tensor, NdArray)
 ```
 
+!!! note
+    Serialization to protobuf formats is not supported for union type hints (unless your union type consists solely of basic types). Please use JSON serialization as an alternative for these cases.
```
Comment (JoanFM) [inline docs/user_guide/sending/serialization.md:307]:
```suggestion
    Serialization to protobuf is not supported for union types involving `BaseDoc` types.
```

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Pull request merged**